### DISCUSSION TITLE
Java version on Mac documentation

### DISCUSSION THREAD

**Discussion by krekr** [ai-republic/bms-to-inverter#65] [General]
I couldn't get the configurator.jar to run, turned out I had a very old (2020) version of Java installed that came with MacOS. If you double-click the .jar in Finder, you'll get a generic error. Using the command line results in:
`Error: LinkageError occurred while loading main class com.airepublic.bmstoinverter.configurator.Configurator
	java.lang.UnsupportedClassVersionError: com/airepublic/bmstoinverter/configurator/Configurator has been compiled by a more recent version of the Java Runtime (class file version 61.0), this version of the Java Runtime only recognizes class file versions up to 59.0`

What made it even more difficult to troubleshoot was that the error referred to the Java runtime (JRE) which you can [download here ](https://www.java.com/en/download/manual.jsp) but doesn't fix the error;

Turns out you need to download the [JDK](https://www.oracle.com/java/technologies/downloads/) (NOT JRE) and link that in your shell:
`/usr/libexec/java_home -V`
`nano ~/.zshrc`
Add:
`export JAVA_HOME=/Library/Java/JavaVirtualMachines/jdk-23.jdk/Contents/Home
#export JAVA_HOME=/Library/Java/JavaVirtualMachines/jdk-15.0.1.jdk/Contents/Home
export PATH=$JAVA_HOME/bin:$PATH`
Reload shell:
`source ~/.zshrc`
Check if new version actually runs:
`java -version  
java version "23" 2024-09-17
Java(TM) SE Runtime Environment (build 23+37-2369)
Java HotSpot(TM) 64-Bit Server VM (build 23+37-2369, mixed mode, sharing)`

Documenting this here for future generations :) Might be something to put in doc.

**Comment by krekr** [comment]
Something similar applies to armhf / armv6 / armv7 for the target; as far as I can tell there is no JDK version for those environments that is new enough to support this software. I was using a Pi Zero W; now switched to Pi4.

**Comment by ai-republic** [comment]
Hi @krekr, 
yes, thanks for this clarification and sharing this. I will update the documentation accordingly 😄 

**_On your comment I actually downgraded the minimum requirements to JDK 8 so that people with a 32-bit microcontroller can use this software!_**

Is recommended though to use a 64bit Java JDK 21+ and a 64bit operating system.
